### ISSUE TITLE
Tracking Issue for raw slice getters (slice_ptr_get)

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#74265]
This is a tracking issue for indexing methods on raw slices.
The feature gate for the issue is `#![feature(slice_ptr_get)]`.

### Steps

- [ ] Merge Implementation
- [ ] Stabilization PR

### Implementation history

* Initial PR: https://github.com/rust-lang/rust/pull/73986

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/issues/74679 is a potential blocker for this.

**Comment by RalfJung** [comment]
With https://github.com/rust-lang/rust/pull/75152, the allocator trait started making use of this.

**Comment by TimDiekmann** [comment]
Especially with `AllocRef` using it, it would come in handy if we could also add

```rust
impl<T> NonNull<[T]> {
    #[inline]
    fn as_mut_ptr(self) -> *mut T {
        self.as_non_null_ptr().as_ptr()
    }
}
```

If we want this I'd make a PR.

**Comment by RalfJung** [comment]
Seems reasonable to me -- @Amanieu what do you think?

**Comment by TimDiekmann** [comment]
An alternative name would be `as_raw_ptr` to avoid confusion with `as_ptr`, which already returns `*mut T`.

**Comment by Amanieu** [comment]
I don't particularly care for the name bikeshed but I agree that this method would be useful.

**Comment by TimDiekmann** [comment]
This issue now also tracks `NonNull::<[T]>::as_mut_ptr()`. I can't edit the OP as I'm not a collaborator of this repository.

**Comment by RalfJung** [comment]
While using these methods for an upcoming PR, I noticed one thing: there is no const-to-mut coercion for raw pointers (at least not for slices, or maybe method lookup just is unable to use it), so one *has* to use `get_unchecked_mut` on `*mut [T]`.

I wonder if it would make sense to just not have the `mut` in the name (but it would likely be inconsistent).

**Comment by steffahn** [comment]
There’s discussion [here](https://internals.rust-lang.org/t/casting-constness-can-be-risky-heres-a-simple-fix/15933) about adding general `*const T` -> `*mut T` and `*mut T` -> `*const T` helper methods (as a way to avoid `as` casts). The method name `as_mut_ptr` sounds like it does exactly that, but it's currently used here for `*mut [T]` -> `*mut T`. Reading the current `NonNull` documentation was also **highly** confusing; particularly when looking at the method list in the side-bar. I would appreciate e.g. some mention to `slice` in the name of the `as_[…]ptr` family of methods.

**Comment by RalfJung** [comment]
These methods are needed to work around the problem described in https://github.com/rust-lang/rust/issues/99437. Though interestingly, to preserve the full semantics of the original code but fix the aliasing problems, we would want bounds-checked getters on raw slices.

**Comment by kornelski** [comment]
@steffahn these methods could be named `cast_mut()` and `cast_shared()` (or is it `const`?), alluding to `ptr::cast()` that is the replacement for the `as` operator. In types like `slice.as_ptr()` and `vec.as_ptr()` this method changes the type, but not shared/mut status.

**Comment by JonathanWoollett-Light** [comment]
Has there been any progress? It appears all required work is done? This feature will be useful in adding safety `munmap`, `madvise` and other `mman` functions in https://crates.io/crates/nix.

**Comment by Ralith** [comment]
The API docs for `get_unchecked` and `get_unchecked_mut` say:

> Calling this method [...] when self is not dereferenceable is [undefined behavior](https://doc.rust-lang.org/reference/behavior-considered-undefined.html)

This is too strong, right? `self` is not dereferenceable when e.g. an exclusive reference to any of its elements exists, but it should be sound to use these methods to construct pointers to other elements, or even to an already-referenced element if the constructed pointer isn't dereferenced. Was the intention to instead specify that the pointer must address allocated memory?

**Comment by RalfJung** [comment]
"dereferenceable" generally means "points to memory that is inbounds of an allocation".

However, these functions take references, so it is indeed UB to call them when there are aliasing conflicts: `get_unchecked` acts like a read of the entire slice, and `get_unchecked_mut` acts like a write of the entire slice. These rules are not final, but until Rust has a precise aliasing model you should follow these conservative rules to ensure maximal chance of being compatible with the future memory model.

**Comment by Ralith** [comment]
> "dereferenceable" generally means "points to memory that is inbounds of an allocation".

That's surprising; in Rust API documentation, of a pointer, I'd expect it to mean "it is legal to convert this pointer into a reference".

> these functions take references

Are we talking about the same functions? `<*const [T]>::get_unchecked` takes a raw pointer by value, and passes it on to `SliceIndex::get_unchecked`:

https://github.com/rust-lang/rust/blob/0ea7ddcc35a2fcaa5da8a7dcfc118c9fb4a63b95/library/core/src/ptr/const_ptr.rs#L1693

which `add`s the index and returns the pointer:

https://github.com/rust-lang/rust/blob/0ea7dd,cc35a2fcaa5da8a7dcfc118c9fb4a63b95/library/core/src/slice/index.rs#L238

I don't see a single reference being accepted, constructed or manipulated at any stage here.

**Comment by RalfJung** [comment]
>      Are we talking about the same functions?

Sorry, I thought you meant `slice::get_unchecked`. I didn't realize which issue this is in :joy: 

> That's surprising; in Rust API documentation, of a pointer, I'd expect it to mean "it is legal to convert this pointer into a reference".

No, that's not what it means.

**Comment by RalfJung** [comment]
I think this has been sitting around long enough, clearly a fix for https://github.com/rust-lang/rust/issues/73987 will not magically materialize. And even with https://github.com/rust-lang/rust/issues/73987 unresolved, as long as we don't let users write arbitrary-self methods on raw pointers, we can still change the autoref behavior later to use raw pointers when that is enough. So while these methods might have a footgun, they are still better than the status quo.

So I propose we move towards stabilizing the raw pointer part of this, and of https://github.com/rust-lang/rust/issues/71146. (I am less sure about the `NonNull` part.)

@rust-lang/types are you aware of anything that would block stabilizing a function that uses a raw pointer self type? (There might already be stable functions that do that, I am not sure.)

**Comment by lcnr** [comment]
https://github.com/rust-lang/rust/issues/71146#issuecomment-1838908803 applies to this feature as well

**Comment by the8472** [comment]
This happened to come up in a libs-api meeting. We want to note that stabilization of this feature will likely be blocked on the `arbitrary_self_types` feature since that will allow moving these methods to

```rust
impl [T] {
   fn as_mut_ptr(*const Self) -> *mut T 
}
```

**Comment by RalfJung** [comment]
Yeah I was about to ask, given @joshtriplett's reply [here](https://github.com/rust-lang/rust/pull/118806#issuecomment-1883421095), what the implications are of that for this feature.

However, I think we should really offer *some* way to work with raw slice pointers. The current situation is pretty unfortunate and nothing has improved for many years. If we can't stabilize this as-is, does @rust-lang/libs-api have any alternatives we could pursue?

**Comment by the8472** [comment]
Well, part of the issue is that similar methods [have been proposed](https://github.com/rust-lang/libs-team/issues/321) for array pointers. We'd either have to duplicate them or see what kind of deref-like coercions self types will provide.

**Comment by RalfJung** [comment]
> or see what kind of deref-like coercions self types will provide

So far, Rust has largely avoided coercions on raw pointers. That's pretty orthogonal to arbitrary-self-types, so I don't expect https://github.com/rust-lang/rfcs/pull/3519 to change that. So another T-lang RFC would be needed to have such coercions on raw ptr methods I think.

**Comment by Rua** [comment]
Is there any interest in implementing a method that *does* perform bounds checking, while still operating on pointers rather than references?

**Comment by RalfJung** [comment]
That would make sense to me, yeah. I am not in t-libs-api though, so this should probably go through an ACP.

**Comment by Rua** [comment]
I ran into an issue with this. `get_unchecked` is implemented for pointers to slices, but not for pointers to fixed-size arrays. We probably want to implement it for arrays too?

**Comment by nazar-pc** [comment]
I found it really strange that both `as_ptr()` and `as_mut_ptr()` return `*mut T`, I'd expect `as_ptr()` to return `*const T`.

This is especially true when trying to keep symmetry with `from_ref()` and `from_mut()`. It doesn't seem right to create an instance with `from_ref()` and then get `*mut T` using `as_ptr()`.

**Comment by RalfJung** [comment]
Are you referring to `NonNull::as_ptr`? It's not clear from your comment. That one is already stable so its name is not up for dispute.

**Comment by nazar-pc** [comment]
I was, but failed to see it was already stabilized long time ago

**Comment by clarfonthey** [comment]
Should `split_at` methods (#95595) be lumped under this feature gate? It seems like the main blocker is the pointer method coercions, and otherwise those methods are just as useful as these. In particular, `split_at` is often used to do const slicing while const traits are still unstable.

I was going to make a PR to flesh out those methods (since they're only for `*mut` pointers, and missing for `*const` and `NonNull`), but wanted to ask this first since I can move the feature gate over in the same PR.
